File: web/src/dtos/Order.ts
import Common from "../utility/Common";
import { CityRes } from "./City";
import { PagedReq } from "./PagedReq";
import { PaymentMethodRes } from "./PaymentMethod";
import { UserRes } from "./User";

export interface OrderRes {
  orderId?: number;
  userId?: string;
  baseSubTotal?: number;
  quantity?: number;
  shippedQuantity?: number;
  orderDate?: Date;
  status?: number;
  paymentMethodId?: number;

  paymentMethod?: PaymentMethodRes;
  user?: UserRes;
  addresses?: OrderAddressRes[];
}

export class OrderReqEdit {
  paymentMethodId?: number;
  addressIdForShipping?: number;
  addressIdForBilling?: number;

  constructor(
    paymentMethodId?: number,
    addressIdForShipping?: number,
    addressIdForBilling?: number
  ) {
    this.paymentMethodId = paymentMethodId;
    this.addressIdForShipping = addressIdForShipping;
    this.addressIdForBilling = addressIdForBilling;
  }
}

export class OrderReqSearch extends PagedReq {
  userId?: string;
  status?: number;

  constructor(
    {
      pageNumber = 1,
      pageSize = Common.DEFAULT_PAGE_SIZE,
      orderBy,
      searchText = "",
    }: PagedReq,
    { userId = "", status = 0 }
  ) {
    super({
      pageNumber: pageNumber,
      pageSize: pageSize,
      orderBy: orderBy,
      searchText: searchText,
    });
    this.userId = userId;
    this.status = status;
  }
}

export enum OrderStatus {
  Pending = 1,
  Confirmed = 2,
  Shipping = 3,
  Delivered = 4,
  Cancelled = 5,
  Complete = 10,
}

export class OrderReqUpdateStatus {
  status?: number = 0;

  constructor(status?: number) {
    this.status = status;
  }
}

export class OrderReqUpdatePaymentMethod {
  paymentMethodId?: number;

  constructor(paymentMethodId?: number) {
    this.paymentMethodId = paymentMethodId;
  }
}

export interface OrderAddressRes {
  orderAddressId?: number;
  firstName?: string;
  lastName?: string;
  phoneNumber?: string;
  line1?: string;
  line2?: string;

  cityId?: number;
  city?: CityRes;

  isShippingAddress?: boolean;
  isBillingAddress?: boolean;
}

File: data/Dtos/Order.cs
using data.Entities;
using Microsoft.EntityFrameworkCore.Metadata.Internal;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using data.Utility.Paging;

namespace data.Dtos
{
    public class OrderRes
    {
        public int OrderId { get; set; }
        public string? UserId { get; set; }
        public UserRes? User { get; set; }
        public decimal BaseSubTotal { get; set; }
        public DateTime OrderDate { get; set; } = DateTime.UtcNow;

        public int PaymentMethodId { get; set; }
        public PaymentMethodRes? PaymentMethod { get; set; }

        public int Status { get; set; }

        // Child tables
        public IEnumerable<OrderAddressRes>? Addresses { get; set; }
        public ICollection<OrderItemRes>? OrderItems { get; set; }
    }

    public class OrderReqSearch : PagedRequestDto
    {
        public string? UserId { get; set; }
        public int Status { get; set; } = 0;
    }

    public class OrderReqEdit
    {
        public int? PaymentMethodId { get; set; }
        [Required]
        public int AddressIdForShipping { get; set; }
        [Required]
        public int AddressIdForBilling { get; set; }
    }

    public class OrderReqUpdateStatus
    {
        [Required]
        public int Status { get; set; }
    }
}
